### PR TITLE
Migrate cluster IDs to ClusterIdentity

### PR DESCRIPTION
**Pull request by CapacitorSet**
# Description

This PR refactors all components to identify clusters by their ClusterIdentity (which internally contains their ID and name) rather than the cluster ID.

This opens the possibility of using user-friendly names in both Kubernetes object names and log messages.

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [x] Unit tests
- [x] E2E tests

### CODE REVIEW

File: pkg/identityManager/iamIdentityProvider.go [hunk 1]
```diff
@@ -78,7 +78,7 @@ func (identityProvider *iamIdentityProvider) ApproveSigningRequest(clusterID,
 
 	// the IAM username has to have <= 64 charaters, we have to take only a prefix from the local clusterID.
 	prefix := identityProvider.localClusterID[:25]
-	username := fmt.Sprintf("%v-%v", prefix, clusterID)
+	username := prefix + cluster.ClusterID
```
Comment (aleoli) [inline pkg/identityManager/iamIdentityProvider.go:81]: why is this change required?

File: pkg/identityManager/iamIdentityProvider.go [hunk 2]
```diff
@@ -98,7 +98,7 @@ func (identityProvider *iamIdentityProvider) ApproveSigningRequest(clusterID,
 		return response, err
 	}
 
-	if err = identityProvider.ensureConfigMap(userArn, clusterID); err != nil {
+	if err = identityProvider.ensureConfigMap(userArn, cluster.ClusterID); err != nil {
```
Comment (aleoli) [inline pkg/identityManager/iamIdentityProvider.go:101]: we should forward the cluster identity

File: pkg/liqoctl/install/aks/aks_test.go [hunk 3]
```diff
@@ -53,7 +53,7 @@ var _ = Describe("Extract elements from AKS", func() {
 		cmd := &cobra.Command{}
 
 		GenerateFlags(cmd)
-		cmd.Flags().String("cluster-name", "", "Name to assign to the Liqo Cluster")
+		cmd.Flags().String("cluster-name", "test-cluster", "Name to assign to the Liqo Cluster")
```
Comment (aleoli) [inline pkg/liqoctl/install/aks/aks_test.go:56]: we should mark this flag as required and not set a default value (the same for the other providers)

File: pkg/tenantNamespace/clusterwide.go [hunk 4]
```diff
@@ -110,10 +111,10 @@ func (nm *tenantNamespaceManager) UnbindOutgoingClusterWideRole(ctx context.Cont
...
-func getClusterRoleName(clusterID string) string {
-	return fmt.Sprintf("%v-%v", clusterRolePrefix, clusterID)
+func getClusterRoleName(cluster discoveryv1alpha1.ClusterIdentity) string {
+	return clusterRolePrefix + "-" + cluster.ClusterID
 }
 
-func getTenantName(clusterID string) string {
-	return fmt.Sprintf("%v-%v", tenantPrefix, clusterID)
+func getTenantName(cluster discoveryv1alpha1.ClusterIdentity) string {
+	return tenantPrefix + "-" + cluster.ClusterID
```
Comment (aleoli) [inline pkg/tenantNamespace/clusterwide.go:119]: we should rollback to the sprintf pattern

File: pkg/tenantNamespace/namespaceManager.go [hunk 5]
```diff
@@ -79,9 +79,9 @@ func (nm *tenantNamespaceManager) CreateNamespace(clusterID string) (ns *v1.Name
 	// exit with an error, and retry during the next iteration.
 	ns = &v1.Namespace{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: strings.Join([]string{tenantNamespaceRoot, clusterID}, "-"),
+			Name: tenantNamespaceRoot + "-" + cluster.ClusterID,
```
Comment (aleoli) [inline pkg/tenantNamespace/namespaceManager.go:82]: the same here (and in the other occurrences)

File: cmd/auth-service/main.go [hunk 6]
```diff
@@ -22,6 +22,7 @@ import (
 	"k8s.io/klog/v2"
 	ctrl "sigs.k8s.io/controller-runtime"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline cmd/auth-service/main.go:25]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: internal/auth-service/identity.go [hunk 7]
```diff
@@ -26,6 +26,7 @@ import (
 	"k8s.io/klog/v2"
 	"k8s.io/utils/trace"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline internal/auth-service/identity.go:29]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/discoverymanager/discovery.go [hunk 8]
```diff
@@ -21,6 +21,8 @@ import (
 
 	"github.com/grandcat/zeroconf"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/discoverymanager/discovery.go:25]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/namespaceMap-controller/create_remote_clients.go [hunk 9]
```diff
@@ -18,28 +18,29 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/klog/v2"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/namespaceMap-controller/create_remote_clients.go:21]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go [hunk 10]
```diff
@@ -28,14 +28,17 @@ import (
 	"k8s.io/klog/v2"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go:31]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/namespaceMap-controller/namespacemap_controller.go [hunk 11]
```diff
@@ -24,6 +24,7 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/namespaceMap-controller/namespacemap_controller.go:27]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/namespaceMap-controller/suite_test.go [hunk 12]
```diff
@@ -27,24 +26,34 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
 	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/namespaceMap-controller/suite_test.go:35]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/namespaceOffloading-controller/namespaceoffloading_controller.go [hunk 13]
```diff
@@ -27,6 +27,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/namespaceOffloading-controller/namespaceoffloading_controller.go:30]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/namespaceOffloading-controller/suite_test.go [hunk 14]
```diff
@@ -33,6 +32,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
 	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/namespaceOffloading-controller/suite_test.go:35]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/resourceoffer-controller/controller_config.go [hunk 15]
```diff
@@ -19,12 +19,13 @@ import (
 
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/resourceoffer-controller/controller_config.go:22]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/liqo-controller-manager/resourceoffer-controller/resourceoffer_controller.go [hunk 16]
```diff
@@ -39,6 +39,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/liqo-controller-manager/resourceoffer-controller/resourceoffer_controller.go:42]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/vkMachinery/forge/creation.go [hunk 17]
```diff
@@ -20,14 +20,15 @@ import (
 	rbacv1 "k8s.io/api/rbac/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/vkMachinery/forge/creation.go:23]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/vkMachinery/forge/forge.go [hunk 18]
```diff
@@ -19,6 +19,7 @@ import (
 
 	v1 "k8s.io/api/core/v1"
 
+	"github.com/liqotech/liqo/apis/discovery/v1alpha1"
```
Comment (aleoli) [inline pkg/vkMachinery/forge/forge.go:22]:
```suggestion
	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
```

File: pkg/tenantNamespace/namespaceManager.go (hunk 5 above)
Comment (CapacitorSet) [reply pkg/tenantNamespace/namespaceManager.go:82]: Why do you think the Sprintf/strings.Join is better?

File: pkg/liqoctl/install/aks/aks_test.go (hunk 3 above)
Comment (CapacitorSet) [reply pkg/liqoctl/install/aks/aks_test.go:56]: This is where values are passed in tests. But you're right, looking at the surrounding code I should have used

```go
Expect(flags.Set("cluster-name", "test-cluster")).To(Succeed())
```

Comment (CapacitorSet) [reply pkg/liqoctl/install/aks/aks_test.go:56]: It is required by ValidateGenericCommandArguments in pkg/liqoctl/install/provider/generic.go, and it must be declared in each test because it is not a custom flag (declared by the provider). However, I will split the declaration and the assignation of the value.

File: pkg/identityManager/iamIdentityProvider.go (hunk 1 above)
Comment (CapacitorSet) [reply pkg/identityManager/iamIdentityProvider.go:81]: My bad, I forgot to add the dash.

File: pkg/tenantNamespace/namespaceManager.go (hunk 5 above)
Comment (aleoli) [reply pkg/tenantNamespace/namespaceManager.go:82]: For two main reasons:
1. it is the approach used in all liqo components
2. it's easier to maintain

You can read in [this post](https://tomasvotruba.com/blog/2018/10/11/hi-my-name-is-tom-conctat-vs-in-sprintf-vs-in-string-variables/) the discussion about different approaches (this is about php, but it is very similar for go)

File: pkg/utils/args/args_test.go [hunk 19]
```diff
@@ -210,6 +211,43 @@ var _ = Describe("ParseArguments", func() {
...
 	})
 
+	Context("ClusterIdentity", func() {
+
+		type parseClusterIdentityTestCase struct {
+			args          []string
+			expectedError OmegaMatcher
+		}
+
+		DescribeTable("CIDR table",
```
Comment (aleoli) [inline pkg/utils/args/args_test.go:221]: The name should be changed

File: pkg/utils/args/args_test.go [hunk 20]
```diff
@@ -210,6 +211,43 @@ var _ = Describe("ParseArguments", func() {
...
+			},
+
+			Entry("no cluster ID", parseClusterIdentityTestCase{
+				args:          []string{"--cluster-name=foo"},
+				expectedError: HaveOccurred(),
+			}),
+			Entry("no cluster name", parseClusterIdentityTestCase{
+				args:          []string{"--cluster-id=foo"},
+				expectedError: HaveOccurred(),
+			}),
```
Comment (aleoli) [inline pkg/utils/args/args_test.go:237]: if the cluster name is required, we have to force the user to set it

1. in the template files the `if .values.discovery.config.clusterName` checks are not required
2. the chart should check if the name is set, and fail accordingly
3. liqoctl should either fail if the name is not set or provide a pseudo-randomic name if none is set

File: pkg/utils/args/cluster-identity.go [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+//
+//   fcFlags := NewClusterIdentityFlags("foreign-cluster", "The %s of the foreign cluster")
+//   flag.Parse()
+//   foreignClusterIdentity := fcFlags.Read()
+func NewClusterIdentityFlags(prefix, description string) ClusterIdentityFlags {
+	return ClusterIdentityFlags{
+		ClusterID:   flag.String(fmt.Sprintf("%s-id", prefix), "", fmt.Sprintf(description, "ID")),
+		ClusterName: flag.String(fmt.Sprintf("%s-name", prefix), "", fmt.Sprintf(description, "name")),
+	}
+}
```
Comment (aleoli) [inline pkg/utils/args/cluster-identity.go:43]: we can define two different functions to create flags for local and for remote clusters, and avoid to pass the description as argument

```go
func NewLocalClusterIdentityFlags
```
and
```go
func NewRemoteClusterIdentityFlags
```

**Review by palexster — APPROVED** [review]

**Review by aleoli — APPROVED** [review]

File: apis/discovery/v1alpha1/foreigncluster_types.go [hunk 22]
```diff
@@ -101,7 +101,7 @@ type ClusterIdentity struct {
 	// Foreign Cluster ID, this is a unique identifier of that cluster.
```
Comment (palexster) [inline apis/discovery/v1alpha1/foreigncluster_types.go:101]: Can you please consolidate your commits?
